File: sdks/csharp/src/Synap.SDK/Modules/StreamManager.cs
using System.Text.Json;
using Synap.SDK.Types;

namespace Synap.SDK.Modules;

/// <summary>
/// Event Stream operations.
/// </summary>
public sealed class StreamManager
{
    private readonly SynapClient _client;

    internal StreamManager(SynapClient client)
    {
        _client = client;
    }

    /// <summary>Creates a new stream room.</summary>
    public async Task CreateRoomAsync(string room, CancellationToken cancellationToken = default)
    {
        using var _ = await _client.SendCommandAsync(
            "stream.create",
            new Dictionary<string, object?> { ["room"] = room },
            cancellationToken).ConfigureAwait(false);
    }

    /// <summary>Deletes a stream room.</summary>
    public async Task DeleteRoomAsync(string room, CancellationToken cancellationToken = default)
    {
        using var _ = await _client.SendCommandAsync(
            "stream.delete",
            new Dictionary<string, object?> { ["room"] = room },
            cancellationToken).ConfigureAwait(false);
    }

    /// <summary>Publishes an event to a stream room.</summary>
    /// <returns>The event offset in the stream.</returns>
    public async Task<long> PublishAsync(
        string room,
        string @event,
        object? data,
        CancellationToken cancellationToken = default)
    {
        using var response = await _client.SendCommandAsync(
            "stream.publish",
            new Dictionary<string, object?>
            {
                ["room"]  = room,
                ["event"] = @event,
                ["data"]  = data,
            },
            cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("offset", out var offset) ? offset.GetInt64() : 0L;
    }

    /// <summary>Reads events from a stream.</summary>
    public async Task<List<StreamEvent>> ReadAsync(
        string room,
        long offset = 0,
        string subscriberId = "sdk-reader",
        CancellationToken cancellationToken = default)
    {
        using var response = await _client.SendCommandAsync(
            "stream.consume",
            new Dictionary<string, object?>
            {
                ["room"]          = room,
                ["subscriber_id"] = subscriberId,
                ["from_offset"]   = offset,
            },
            cancellationToken).ConfigureAwait(false);

        if (!response.RootElement.TryGetProperty("events", out var events) ||
            events.ValueKind != JsonValueKind.Array)
        {
            return new List<StreamEvent>();
        }

        var result = new List<StreamEvent>();
        foreach (var eventElement in events.EnumerateArray())
        {
            result.Add(new StreamEvent
            {
                Offset    = eventElement.TryGetProperty("offset", out var off) ? off.GetInt64() : 0L,
                Event     = eventElement.TryGetProperty("event", out var evt) ? evt.GetString() ?? string.Empty : string.Empty,
                Data      = eventElement.TryGetProperty("data", out var dat) ? dat : default,
                Timestamp = eventElement.TryGetProperty("timestamp", out var ts) ? ts.GetInt64() : 0L,
                Room      = eventElement.TryGetProperty("room", out var rm) ? rm.GetString() : null,
            });
        }

        return result;
    }

    /// <summary>Lists all stream rooms.</summary>
    public async Task<List<string>> ListRoomsAsync(CancellationToken cancellationToken = default)
    {
        using var response = await _client.SendCommandAsync(
            "stream.list",
            new Dictionary<string, object?>(),
            cancellationToken).ConfigureAwait(false);

        if (response.RootElement.TryGetProperty("rooms", out var rooms) && rooms.ValueKind == JsonValueKind.Array)
        {
            var result = new List<string>();
            foreach (var room in rooms.EnumerateArray())
            {
                var s = room.GetString();
                if (s is not null)
                {
                    result.Add(s);
                }
            }

            return result;
        }

        return new List<string>();
    }

    /// <summary>Gets stream statistics.</summary>
    public async Task<Dictionary<string, JsonElement>> StatsAsync(
        string room,
        CancellationToken cancellationToken = default)
    {
        using var response = await _client.ExecuteAsync("stream.stats", room, null, cancellationToken).ConfigureAwait(false);

        var result = new Dictionary<string, JsonElement>();
        foreach (var property in response.RootElement.EnumerateObject())
        {
            result[property.Name] = property.Value;
        }

        return result;
    }
}

File: sdks/java/src/main/java/com/hivellm/synap/StreamManager.java
package com.hivellm.synap;

import com.fasterxml.jackson.databind.JsonNode;
import com.hivellm.synap.types.Event;
import com.hivellm.synap.types.StreamStats;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Event stream operations.
 *
 * <p>Obtain an instance via {@link SynapClient#stream()}.
 */
public final class StreamManager {

    private final SynapClient client;

    StreamManager(SynapClient client) {
        this.client = client;
    }

    /**
     * Creates a new stream room.
     *
     * @param room      room name
     * @param maxEvents maximum number of events the room retains (0 = unlimited)
     * @throws SynapException on network or server error
     */
    public void create(String room, long maxEvents) {
        Map<String, Object> payload = SynapClient.newPayload();
        payload.put("room", room);
        payload.put("max_events", maxEvents);
        client.sendCommand("stream.create", payload);
    }

    /**
     * Publishes an event to the named stream room.
     *
     * @param room      room name
     * @param eventType event type label (e.g. {@code "message"}, {@code "click"})
     * @param data      event data (any JSON-serializable object)
     * @return the stream offset assigned to this event
     * @throws SynapException on network or server error
     */
    public long publish(String room, String eventType, Object data) {
        Map<String, Object> payload = SynapClient.newPayload();
        payload.put("room", room);
        payload.put("event", eventType);
        payload.put("data", data);

        JsonNode responsePayload = client.sendCommand("stream.publish", payload);

        JsonNode offsetNode = responsePayload.get("offset");
        return offsetNode != null ? offsetNode.asLong(0L) : 0L;
    }

    /**
     * Consumes events from a stream starting at the given offset.
     *
     * @param room   room name
     * @param offset starting offset (inclusive); use 0 to read from the beginning
     * @param limit  maximum number of events to return
     * @return list of events (never null; may be empty)
     * @throws SynapException on network or server error
     */
    public List<Event> consume(String room, long offset, int limit) {
        Map<String, Object> payload = SynapClient.newPayload();
        payload.put("room", room);
        payload.put("subscriber_id", "sdk-default");
        payload.put("from_offset", offset);
        payload.put("limit", limit);

        JsonNode responsePayload = client.sendCommand("stream.consume", payload);

        List<Event> result = new ArrayList<>();
        JsonNode eventsNode = responsePayload.get("events");
        if (eventsNode != null && eventsNode.isArray()) {
            for (JsonNode eventNode : eventsNode) {
                try {
                    result.add(client.mapper.treeToValue(eventNode, Event.class));
                } catch (Exception e) {
                    throw SynapException.invalidResponse("Failed to deserialize Event: " + e.getMessage());
                }
            }
        }
        return result;
    }

    /**
     * Returns statistics for the named stream room.
     *
     * @param room room name
     * @return a {@link StreamStats} snapshot
     * @throws SynapException on network or server error
     */
    public StreamStats stats(String room) {
        Map<String, Object> payload = SynapClient.newPayload();
        payload.put("room", room);
        JsonNode responsePayload = client.sendCommand("stream.stats", payload);

        try {
            return client.mapper.treeToValue(responsePayload, StreamStats.class);
        } catch (Exception e) {
            throw SynapException.invalidResponse("Failed to deserialize StreamStats: " + e.getMessage());
        }
    }

    /**
     * Lists the names of all stream rooms currently registered on the server.
     *
     * @return list of room names (never null; may be empty)
     * @throws SynapException on network or server error
     */
    public List<String> list() {
        JsonNode responsePayload = client.sendCommand("stream.list");

        List<String> result = new ArrayList<>();
        JsonNode roomsNode = responsePayload.get("rooms");
        if (roomsNode != null && roomsNode.isArray()) {
            for (JsonNode item : roomsNode) {
                if (!item.isNull()) {
                    result.add(item.asText());
                }
            }
        }
        return result;
    }

    /**
     * Deletes the named stream room and all its retained events.
     *
     * @param room room name
     * @throws SynapException on network or server error
     */
    public void delete(String room) {
        Map<String, Object> payload = SynapClient.newPayload();
        payload.put("room", room);
        client.sendCommand("stream.delete", payload);
    }
}
